modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/connectivity/oauth/OAuthConnectionProviderWrapper.java

```diff
@@ -0,0 +1,114 @@
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.extension.internal.runtime.connectivity.oauth;
+
+import static java.lang.String.format;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.stream.Collectors.toList;
+import static org.mule.runtime.module.extension.internal.runtime.connectivity.oauth.ExtensionsOAuthUtils.toAuthorizationCodeState;
+import static org.mule.runtime.module.extension.internal.util.IntrospectionUtils.getFields;
+import org.mule.runtime.api.connection.ConnectionException;
+import org.mule.runtime.api.connection.ConnectionProvider;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.core.api.retry.RetryPolicyTemplate;
+import org.mule.runtime.core.internal.connection.ReconnectableConnectionProviderWrapper;
+import org.mule.runtime.core.util.func.Once;
+import org.mule.runtime.extension.api.annotation.connectivity.oauth.OAuthCallbackValue;
+import org.mule.runtime.extension.api.connectivity.oauth.AuthorizationCodeState;
+import org.mule.runtime.extension.api.exception.IllegalConnectionProviderModelDefinitionException;
+import org.mule.runtime.module.extension.internal.runtime.connectivity.NoConnectivityTest;
+import org.mule.runtime.module.extension.internal.util.FieldSetter;
+import org.mule.runtime.oauth.api.state.ResourceOwnerOAuthContext;
+
+import java.lang.reflect.Field;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A {@link ReconnectableConnectionProviderWrapper} which makes sure that by the time the
+ * {@link ConnectionProvider#connect()} method is invoked on the delegate, the authorization dance has
+ * been completed and the {@link AuthorizationCodeState} and {@link OAuthCallbackValue} fields have
+ * been properly injected
+ *
+ * @since 4.0
+ */
+public class OAuthConnectionProviderWrapper<C> extends ReconnectableConnectionProviderWrapper<C> implements NoConnectivityTest {
+
+  private final OAuthConfig oauthConfig;
+  private final Map<Field, String> callbackValues;
+  private final ExtensionsOAuthManager oauthManager;
+  private final FieldSetter<ConnectionProvider<C>, AuthorizationCodeState> authCodeStateSetter;
+  private final Once dance;
+
+  public OAuthConnectionProviderWrapper(ConnectionProvider<C> delegate,
+                                        OAuthConfig oauthConfig,
+                                        Map<Field, String> callbackValues,
+                                        ExtensionsOAuthManager oauthManager,
+                                        boolean disableValidation,
+                                        RetryPolicyTemplate retryPolicyTemplate) {
+    super(delegate, disableValidation, retryPolicyTemplate);
+    this.oauthConfig = oauthConfig;
+    this.oauthManager = oauthManager;
+    authCodeStateSetter = getAuthCodeStateSetter(delegate);
+    this.callbackValues = unmodifiableMap(callbackValues);
+    dance = Once.of(this::updateAuthState);
```

Comment:
it's not about failing, it's about not doing the same thing on each connect invokation